src/function.ts

```diff
@@ -656,24 +700,42 @@ export class Function<
             ],
     });
 
-    super(_resource);
-
-    // retrieve and bind all found native integrations. Will fail if the integration does not support native integration.
-    const nativeIntegrationsPrewarm = func.integrations.flatMap(
-      ({ integration, args }) => {
-        const integ = new IntegrationImpl(integration).native;
-        integ.bind(this, args);
-        return integ.preWarm ? [integ.preWarm] : [];
-      }
-    );
-
     // Start serializing process, add the callback to the promises so we can later ensure completion
     Function.promises.push(
-      callbackLambdaCode.generate(nativeIntegrationsPrewarm)
+      (async () => {
+        try {
+          await callbackLambdaCode.generate(nativeIntegrationsPrewarm);
+        } catch (e) {
+          if (e instanceof SynthError) {
+            throw new SynthError(
+              e.code,
+              `While serializing ${_resource.node.path}:\n\n${e.message}`
+            );
+          } else if (e instanceof Error) {
+            throw Error(
+              `While serializing ${_resource.node.path}:\n\n${e.message}`
+            );
+          }
+        }
+        return;
+      })()
     );
   }
 }
 
+function getInvokedIntegrations(ast: FunctionDecl): IntegrationCallExpr[] {
+  const nodes: IntegrationCallExpr[] = [];
+  visitEachChild(ast, function visit(node: FunctionlessNode): FunctionlessNode {
+    if (isIntegrationCallExpr(node)) {
+      nodes.push(node);
+    }
```

Comment:
Here, a ref() could be another `function`. We will need to get its closure and recurse in. That's a change I was working on with the `global.wrapClosure` and `getClosure(func): FunctionDecl | undefined` .

Reply:
yeah, ha. this worked for the current state.

Reply:
No I'm really happy for this change. Saves me doing it. This is in the right state now